akka-http/src/main/scala/akka/http/scaladsl/server/PathMatcher.scala

```diff
@@ -80,13 +80,13 @@ abstract class PathMatcher[L](implicit val ev: Tuple[L]) extends (Path ⇒ PathM
     new PathMatcher[lift.Out]()(lift.OutIsTuple) {
       require(min >= 0, "`min` must be >= 0")
       require(max >= min, "`max` must be >= `min`")
-      def apply(path: Path) = rec(path, 1)
+      def apply(path: Path) = rec(path, 0)
       def rec(path: Path, count: Int): Matching[lift.Out] = {
```

Comment:
This piece of code is very hard to reason about. So to benefit the reviewer, I am explaining the change in words.

`rec(path: Path, count: Int)` is recursive path matcher, in which path corresponds to the provided/remaining path and count corresponds to number of times the expected pattern appears in the path.

so, `count` should start with 0, instead of 1. (which is the root cause for the bug)

Reply:
Maybe rename `count` to `alreadyFound`?

Thanks, good explanation. I guess one of the problems of the complexity here is matching the separator between the elements (as so often when you try to do that). Do you match the separator first (if it's not the first element) and then the element? Or do you match an element first and then try to see if there's a separator and only go on in that case.

I tried turning that logic around (only matching separator if it's not the first element), wdyt about that:

```scala
def apply(path: Path) = matchNext(path, 0)
def matchNext(path: Path, alreadyFound: Int): Matching[lift.Out] = {
  def done = if (alreadyFound >= min) Matched(path, lift()) else Unmatched

  def matchSeparatorIfNeeded(path: Path): Matching[Unit] =
    if (alreadyFound >= 1) separator(path) else Matched(path, ())
  def matchElement(start: Path): Matching[lift.Out] =
    self(start)
      .andThen { (remaining, extractions) ⇒
        matchNext(remaining, alreadyFound + 1)
          .map(result ⇒ lift(extractions, result))
      }
      .orElse(done)

  if (alreadyFound < max)
    matchSeparatorIfNeeded(path)
      .andThen { (remaining, _) ⇒ matchElement(remaining) }
      .orElse(done)
  else done
}
```